File: latest/week1/workshop2_m10/W02_M10_demo.py
"""W02-M10: Lab — Building a Visual Report — Python — UNYT Tirana"""
import numpy as np, pandas as pd, matplotlib.pyplot as plt, os
from matplotlib.gridspec import GridSpec
os.makedirs("output", exist_ok=True); np.random.seed(42)

apps = pd.read_csv("googleplaystore.csv")
apps_clean = (apps.query("Type in ['Free','Paid']")
    .assign(Reviews=lambda d: pd.to_numeric(d["Reviews"], errors="coerce"),
            Last_Updated=lambda d: pd.to_datetime(d["Last Updated"], format="%B %d, %Y", errors="coerce"))
    .assign(Year=lambda d: d["Last_Updated"].dt.year).copy())

# Build 6-panel dashboard
fig = plt.figure(figsize=(14, 9))
gs = GridSpec(2, 3, figure=fig, hspace=0.4, wspace=0.35)

# (a) Bar
ax = fig.add_subplot(gs[0, 0])
top8 = apps_clean.value_counts("Category").head(8).sort_values()
colors = ["#BBBBBB"] * 8; colors[-1] = "#1565C0"
ax.barh(top8.index, top8.values, color=colors, height=0.5, edgecolor="none")
for i, v in enumerate(top8.values):
    ax.text(v + 15, i, f"{v:,}", va="center", fontsize=6, fontweight="bold",
            color="#1565C0" if i == 7 else "#888")
ax.set_title("(a) Top Categories", fontsize=9, fontweight="bold")
for sp in ["top","right","left"]: ax.spines[sp].set_visible(False)
ax.set_xticks([]); ax.tick_params(left=False, labelsize=6)

# (b) Histogram
ax = fig.add_subplot(gs[0, 1])
ratings = apps_clean["Rating"].dropna()
ax.hist(ratings, bins=30, color="#1565C0", edgecolor="white", linewidth=0.3)
ax.axvline(x=ratings.mean(), color="#E53935", linewidth=1.5, linestyle="--")
ax.text(ratings.mean() + 0.05, ax.get_ylim()[1] * 0.9, f"Mean: {ratings.mean():.2f}",
        fontsize=7, color="#E53935", fontweight="bold")
ax.set_title("(b) Rating Distribution", fontsize=9, fontweight="bold")
ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False); ax.tick_params(labelsize=6)

# (c) Scatter
ax = fig.add_subplot(gs[0, 2])
df_sc = apps_clean.dropna(subset=["Rating"]).query("Reviews > 0")
for t, c in [("Free", "#1565C0"), ("Paid", "#E53935")]:
    m = df_sc["Type"] == t
    ax.scatter(df_sc.loc[m, "Reviews"], df_sc.loc[m, "Rating"], s=5, c=c, alpha=0.2, edgecolors="none", label=t)
ax.set_xscale("log"); ax.legend(fontsize=5, title="Type", title_fontsize=6)
ax.set_title("(c) Reviews vs Rating", fontsize=9, fontweight="bold")
ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False); ax.tick_params(labelsize=6)

# (d) Boxplot
ax = fig.add_subplot(gs[1, 0])
free_r = apps_clean.loc[apps_clean["Type"] == "Free", "Rating"].dropna()
paid_r = apps_clean.loc[apps_clean["Type"] == "Paid", "Rating"].dropna()
bp = ax.boxplot([free_r, paid_r], labels=["Free", "Paid"], patch_artist=True, widths=0.4, notch=True)
bp["boxes"][0].set_facecolor("#BBDEFB"); bp["boxes"][1].set_facecolor("#FFCDD2")
ax.scatter([1, 2], [free_r.mean(), paid_r.mean()], c="#E53935", s=40, zorder=5, marker="D", edgecolors="white")
ax.set_title("(d) Rating by Type", fontsize=9, fontweight="bold")
ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False); ax.tick_params(labelsize=6)

# (e) Line
ax = fig.add_subplot(gs[1, 1])
yearly = apps_clean.query("2010 <= Year <= 2018").groupby(["Year", "Type"]).size().reset_index(name="n")
for t, c in [("Free", "#1565C0"), ("Paid", "#E53935")]:
    sub = yearly.query("Type == @t")
    ax.plot(sub["Year"], sub["n"], "o-", color=c, linewidth=1.5, markersize=4)
    ax.text(sub["Year"].iloc[-1] + 0.2, sub["n"].iloc[-1], t, fontsize=6, fontweight="bold", color=c, va="center")
ax.set_xlim(2009.5, 2019); ax.set_title("(e) Apps per Year", fontsize=9, fontweight="bold")
ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False); ax.tick_params(labelsize=6)

# (f) Stacked bar
ax = fig.add_subplot(gs[1, 2])
keep_cr = ["Everyone", "Teen", "Mature 17+", "Everyone 10+"]
ct = pd.crosstab(apps_clean.loc[apps_clean["Content Rating"].isin(keep_cr), "Content Rating"], apps_clean["Type"])
ct.plot.bar(stacked=True, color=["#1565C0", "#E53935"], edgecolor="white", ax=ax, legend=True)
ax.set_title("(f) Content × Type", fontsize=9, fontweight="bold")
ax.tick_params(axis="x", rotation=30, labelsize=5); ax.legend(fontsize=5)
ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)

fig.suptitle("Google Play Store: Complete Visual Report\nWorkshop 2 Lab — Grammar of Graphics Applied",
             fontsize=13, fontweight="bold")
fig.text(0.5, 0.01, "Source: Kaggle | UNYT Data Visualization Course | Prof.Asoc. Endri Raco",
         ha="center", fontsize=8, fontstyle="italic", color="#888")
fig.savefig("output/W02_M10_report.pdf", bbox_inches="tight")
fig.savefig("output/W02_M10_report.png", dpi=300, bbox_inches="tight")
plt.close()
print("Workshop 2 Lab Complete (Python)")

File: latest/week1/workshop2_m10/W02_M10_demo.R
# ============================================================
# Workshop 2 — Module 10: Lab — Building a Visual Report
# R Demonstration Script — UNYT Tirana
# ============================================================
library(tidyverse)
library(patchwork)
library(scales)
dir.create("output", showWarnings = FALSE)

apps <- read_csv("googleplaystore.csv") |>
  filter(Type %in% c("Free", "Paid")) |>
  mutate(Reviews = as.numeric(Reviews),
         Last_Updated = mdy(`Last Updated`),
         Year = year(Last_Updated))

# ── (a) BAR: Top 8 categories ────────────────────────────
p_bar <- apps |> count(Category, sort = TRUE) |> slice_max(n, n = 8) |>
  mutate(Category = fct_reorder(Category, n), hl = Category == "FAMILY") |>
  ggplot(aes(x = Category, y = n, fill = hl)) +
  geom_col(width = 0.5, show.legend = FALSE) +
  geom_text(aes(label = comma(n)), hjust = -0.15, size = 3, fontface = "bold") +
  scale_fill_manual(values = c("TRUE" = "#1565C0", "FALSE" = "#BBBBBB")) +
  coord_flip() + theme_minimal() +
  theme(panel.grid.major.y = element_blank()) +
  labs(title = "(a) FAMILY Leads with 1,832 Apps", x = NULL, y = NULL)

# ── (b) HISTOGRAM: Rating distribution ───────────────────
mean_r <- mean(apps$Rating, na.rm = TRUE)
p_hist <- ggplot(apps, aes(x = Rating)) +
  geom_histogram(bins = 30, fill = "#1565C0", color = "white", linewidth = 0.3) +
  geom_vline(xintercept = mean_r, color = "#E53935", linewidth = 1, linetype = "dashed") +
  annotate("text", x = mean_r + 0.1, y = Inf, vjust = 1.5,
           label = paste0("Mean: ", round(mean_r, 2)), color = "#E53935",
           fontface = "bold", size = 3) +
  theme_minimal() + labs(title = "(b) Rating Distribution", x = "Rating", y = "Count")

# ── (c) SCATTER: Reviews vs Rating ───────────────────────
p_scatter <- apps |> filter(!is.na(Rating), Reviews > 0) |>
  ggplot(aes(x = Reviews, y = Rating, color = Type)) +
  geom_point(alpha = 0.3, size = 0.8) +
  scale_x_log10(labels = comma) +
  scale_color_manual(values = c(Free = "#1565C0", Paid = "#E53935")) +
  theme_minimal() + labs(title = "(c) Reviews vs Rating", x = "Reviews (log)", y = "Rating")

# ── (d) BOXPLOT: Rating by Type ──────────────────────────
p_box <- apps |> filter(!is.na(Rating)) |>
  ggplot(aes(x = Type, y = Rating, fill = Type)) +
  geom_boxplot(width = 0.5, notch = TRUE, outlier.alpha = 0.1) +
  stat_summary(fun = mean, geom = "point", shape = 18, size = 3, color = "#E53935") +
  scale_fill_manual(values = c(Free = "#BBDEFB", Paid = "#FFCDD2")) +
  theme_minimal() + theme(legend.position = "none") +
  labs(title = "(d) Rating by Type", y = "Rating", x = NULL)

# ── (e) LINE: Apps per year ──────────────────────────────
p_line <- apps |> filter(Year >= 2010, Year <= 2018) |> count(Year, Type) |>
  ggplot(aes(x = Year, y = n, color = Type)) +
  geom_line(linewidth = 1.2) + geom_point(size = 2) +
  scale_color_manual(values = c(Free = "#1565C0", Paid = "#E53935")) +
  theme_minimal() + theme(legend.position = "bottom") +
  labs(title = "(e) Apps Added per Year", x = "Year", y = "Count")

# ── (f) STACKED: Content Rating × Type ───────────────────
p_stacked <- apps |>
  filter(`Content Rating` %in% c("Everyone","Teen","Mature 17+","Everyone 10+")) |>
  ggplot(aes(x = `Content Rating`, fill = Type)) +
  geom_bar(position = "stack") +
  scale_fill_manual(values = c(Free = "#1565C0", Paid = "#E53935")) +
  theme_minimal() + theme(axis.text.x = element_text(angle = 30, hjust = 1, size = 7)) +
  labs(title = "(f) Content Rating × Type", x = NULL, y = "Count")

# ── COMPOSE DASHBOARD ────────────────────────────────────
dashboard <- (p_bar | p_hist | p_scatter) / (p_box | p_line | p_stacked) +
  plot_annotation(
    title = "Google Play Store: Complete Visual Report",
    subtitle = "Workshop 2 Lab — Grammar of Graphics Applied",
    caption = "Source: Kaggle | UNYT Data Visualization Course | Prof.Asoc. Endri Raco"
  )

ggsave("output/W02_M10_report.pdf", dashboard, width = 14, height = 9, device = cairo_pdf)
ggsave("output/W02_M10_report.png", dashboard, width = 14, height = 9, dpi = 300)
cat("\n── Workshop 2 Lab Complete (R) ──\n")
